File: analysis/import/SCMLogParser/src/test/java/de/maibornwolff/codecharta/importer/scmlogparser/converter/ProjectConverterTest.java
package de.maibornwolff.codecharta.importer.scmlogparser.converter;

import de.maibornwolff.codecharta.importer.scmlogparser.converter.ProjectConverter;
import de.maibornwolff.codecharta.importer.scmlogparser.input.Commit;
import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification;
import de.maibornwolff.codecharta.importer.scmlogparser.input.VersionControlledFile;
import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.MetricsFactory;
import de.maibornwolff.codecharta.model.Project;
import org.junit.Test;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;

public class ProjectConverterTest {

    private final MetricsFactory metricsFactory = mock(MetricsFactory.class);

    private static List<Modification> modificationsByFilename(String... filenames) {
        return Stream.of(filenames).map(Modification::new).collect(Collectors.toList());
    }

    @Test
    public void canCreateAnEmptyProject() throws Exception {
        // given
        ProjectConverter projectConverter = new ProjectConverter(true);
        String projectname = "Projectname";

        // when
        Project project = projectConverter.convert(projectname, Collections.emptyList());

        //then
        assertThat(project.getNodes()).hasSize(0);
        assertThat(project.getProjectName()).isEqualTo(projectname);
    }

    @Test
    public void canConvertProjectWithAuthors() {
        //given
        ProjectConverter projectConverter = new ProjectConverter(true);
        String projectname = "ProjectWithAuthors";
        VersionControlledFile file1 = new VersionControlledFile("File 1", metricsFactory);
        file1.registerCommit(new Commit("Author", modificationsByFilename("File 1, File 2"), LocalDateTime.now()));

        //when
        Project project = projectConverter.convert(projectname, Arrays.asList(file1));

        //then
        assertThat(project.getRootNode().getChildren().get(0).getAttributes().containsKey("authors")).isTrue();
    }

    @Test
    public void canConvertProjectWithoutAuthors() {
        //given
        ProjectConverter projectConverter = new ProjectConverter(false);
        String projectname = "ProjectWithoutAuthors";
        VersionControlledFile file1 = new VersionControlledFile("File 1", metricsFactory);
        file1.registerCommit(new Commit("Author", modificationsByFilename("File 1, File 2"), LocalDateTime.now()));

        //when
        Project project = projectConverter.convert(projectname, Arrays.asList(file1));

        //then
        assertThat(project.getRootNode().getChildren().get(0).getAttributes().containsKey("authors")).isFalse();
    }


}

File: analysis/import/SCMLogParser/src/test/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/converter/ProjectConverterTest.kt
package de.maibornwolff.codecharta.importer.scmlogparser.converter

import de.maibornwolff.codecharta.importer.scmlogparser.input.Commit
import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification
import de.maibornwolff.codecharta.importer.scmlogparser.input.VersionControlledFile
import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.MetricsFactory
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import java.time.OffsetDateTime
import java.util.*

class ProjectConverterTest {

    private val metricsFactory = mockk<MetricsFactory>()

    @Before
    fun setup() {
        every { metricsFactory.createMetrics() } returns listOf()
    }

    private fun modificationsByFilename(vararg filenames: String): List<Modification> {
        return filenames.map { Modification(it) }
    }

    @Test
    @Throws(Exception::class)
    fun canCreateAnEmptyProject() {
        // given
        val projectConverter = ProjectConverter(true)

        // when
        val project = projectConverter.convert(emptyList(), metricsFactory)

        //then
        assertThat(project.rootNode.leaves).hasSize(1)
    }

    @Test
    fun canConvertProjectWithAuthors() {
        //given
        val projectConverter = ProjectConverter(true)
        val file1 = VersionControlledFile("File 1", metricsFactory)
        file1.registerCommit(Commit("Author", modificationsByFilename("File 1", "File 2"), OffsetDateTime.now()))

        //when
        val project = projectConverter.convert(Arrays.asList(file1), metricsFactory)

        //then
        assertThat(project.rootNode.children[0].attributes.containsKey("authors")).isTrue()
    }

    @Test
    fun canConvertProjectWithoutAuthors() {
        //given
        val projectConverter = ProjectConverter(false)
        val file1 = VersionControlledFile("File 1", metricsFactory)
        file1.registerCommit(Commit("Author", modificationsByFilename("File 1", "File 2"), OffsetDateTime.now()))

        //when
        val project = projectConverter.convert(Arrays.asList(file1), metricsFactory)

        //then
        assertThat(project.rootNode.children[0].attributes.containsKey("authors")).isFalse()
    }

    @Test
    fun edgesAreRegisteredInProject() {
        //given
        val projectConverter = ProjectConverter(true)
        val metricsFactory = MetricsFactory().createMetrics()
        val file1 = VersionControlledFile("File 1", metricsFactory)
        val commit = Commit("Author", modificationsByFilename("File 1", "File 2"), OffsetDateTime.now())
        for (i in 0..4) {
            file1.registerCommit(commit)
        }

        //when
        val project = projectConverter.convert(listOf(file1), MetricsFactory())

        //then
        assertThat(project.edges.size).isEqualTo(1)
        assertThat(project.edges[0].toNodeName).isEqualTo("/root/File 2")
        assertThat(project.edges[0].fromNodeName).isEqualTo("/root/File 1")
    }

    @Test
    fun attributeTypesAreCreated() {
        val projectConverter = ProjectConverter(false)
        val project = projectConverter.convert(listOf(), MetricsFactory())

        assertThat(project.attributeTypes).containsKeys("edges", "nodes")
    }
}
